Adding external dialects (such as STorM32.xml or OpenHD.xml)

olliw42 opened the pull request:
hey folks

this PR needs some introducionary comments. I am following, as far as it's possible through the web, the current discussion & efforts regarding the MAVLink governance policy, and few directions which seem to be taken are IMHO not particularly smart. Maybe not the main, but one point is related to the topic of dialects. And that's why I'm doing this PR here, in order to expose it, and I will argue from two perspectives, broad and general, and practical use-case based.

I'd like to start with saying that I certainly fully understand that the dialects which should be admitted to the MAVLink standard need some strong regulation, and that they need to be backed by strong stakeholders. No argument here. However, that's going to be my main point, dialects could also be admitted to the MAVLink repository on a lesser level of regulation. Note the distinction between standard and repository I'm trying to establish here.

From a general perspective I argue that otherwise the MAVLink world may lose key intellectual resources, specifically hobbyists as well as university students, and this sort of folks. I think everyone here agrees that this pool of resources has driven the story so much in the past, and does still today. I think it is, e.g., fair to say that BetaFlight had tremendous impactand I think it is probably also correct to say that PX4 strongly has benefited and still benefits from close ties to university. Clearly, MAVLink should strive to become an industry standard. If MAVLink is however governed in that way, to say it somewhat pointedly, in a way there only multi-billion dollar companies have a say, this backbone is obviously going to be lost. I thus raise my voice here not a as a stakeholder or multi-billion dollar company, but as member of that diffuse pool of resources I was trying to describe.

The specific motivation for the PR comes from this use case: For whatever good or bad reasons I've decided to draft a dialect storm32.xml, which includes things for operating a STorM32 gimbal and its camera MAVLink support and things like quick shots. At the same time OpenHD, a well-recognized open source FPV wireless transmission system which also permits establishing a MAVLink link, has also drafted its MAVLink dialect. Obviously, it would be highly interesting to make it possible to use OpenHD and STorM32 together in order to achieve things like what DJI and others offer since long, but for free. Obviously also however, some coordination is necessary in order to avoid message and command ID clashes.

While for that specific use case this coordination could certainly easily be done outside of MAVLink, I argue that this is not to the benefit of the community. It would be much better if there would be a way for anyone to see what range of IDs are occupied or claimed by various external projects. This does not requite these dialects to become part of the standard, but only to be visible in a central place, which would allow "significant" project - where it is not clear what "significant" is exactly supposed to mean - to set their ID ranges such as to not conflict with either the standard or other "significant" projects.

That is, what I'm proposing is a way for "significant" projects to submit their dialects to a part of the MAVLink repository which is outside of the standard, but maintained to a level that ID clashes are avoided, and possibly some basic style, language, and such things. I believe that the available message and command ID range allows for this.

And I'd like to get the storm32.xml dialect being admitted to that part of the repository.


Cheers, Olli

LorenzMeier commented:
We are currently phasing out dialects that have had very few developers / contributors involved in other cases, so adding this will be, I'm afraid, a hard sell.

auturgy commented:
I agree with your arguments - getting the balance right between assuring accessibility and avoiding noise/dead wood is a challenge.  You’re an active contributor to mavlink development, so I’m very much in support of your approach (although I do wish ArduPilot could work with you a bit better, to remove your need to maintain a fork).
A couple of points though.
1. I’m not sure an absolute reference to include a local file from your pc is intended (line 13)?
I think we should try to rectify the underlying issue that is driving that requirement, rather than work around it, or at least use a reference that is user accessible (GitHub?).
2. Where messages are WIP (or Very WIP), it may be worth using the WIP xml tag. Noting your desire to manage this dialect external to the “standard”, I’m not hard over on this.

Definitely some good discussion triggers. I’ve seen your QuickShot YouTube video - really nice!

Regards,

James 

> On 8 Dec 2020, at 7:40 pm, olliw42 <notifications@github.com> wrote:
> 
> ﻿
> hey folks
> 
> this PR needs some introducionary comments. I am following, as far as it's possible through the web, the current discussion & efforts regarding the MAVLink governance policy, and few directions which seem to be taken are IMHO not particularly smart. Maybe not the main, but one point is related to the topic of dialects. And that's why I'm doing this PR here, in order to expose it, and I will argue from two perspectives, broad and general, and practical use-case based.
> 
> I'd like to start with saying that I certainly fully understand that the dialects which should be admitted to the MAVLink standard need some strong regulation, and that they need to be backed by strong stakeholders. No argument here. However, that's going to be my main point, dialects could also be admitted to the MAVLink repository on a lesser level of regulation. Note the distinction between standard and repository I'm trying to establish here.
> 
> From a general perspective I argue that otherwise the MAVLink world may lose key intellectual resources, specifically hobbyists as well as university students, and this sort of folks. I think everyone here agrees that this pool of resources has driven the story so much in the past, and does still today. I think it is, e.g., fair to say that BetaFlight had tremendous impactand I think it is probably also correct to say that PX4 strongly has benefited and still benefits from close ties to university. Clearly, MAVLink should strive to become an industry standard. If MAVLink is however governed in that way, to say it somewhat pointedly, in a way there only multi-billion dollar companies have a say, this backbone is obviously going to be lost. I thus raise my voice here not a as a stakeholder or multi-billion dollar company, but as member of that diffuse pool of resources I was trying to describe.
> 
> The specific motivation for the PR comes from this use case: For whatever good or bad reasons I've decided to draft a dialect storm32.xml, which includes things for operating a STorM32 gimbal and its camera MAVLink support and things like quick shots. At the same time OpenHD, a well-recognized open source FPV wireless transmission system which also permits establishing a MAVLink link, has also drafted its MAVLink dialect. Obviously, it would be highly interesting to make it possible to use OpenHD and STorM32 together in order to achieve things like what DJI and others offer since long, but for free. Obviously also however, some coordination is necessary in order to avoid message and command ID clashes.
> 
> While for that specific use case this coordination could certainly easily be done outside of MAVLink, I argue that this is not to the benefit of the community. It would be much better if there would be a way for anyone to see what range of IDs are occupied or claimed by various external projects. This does not requite these dialects to become part of the standard, but only to be visible in a central place, which would allow "significant" project - where it is not clear what "significant" is exactly supposed to mean - to set their ID ranges such as to not conflict with either the standard or other "significant" projects.
> 
> That is, what I'm proposing is a way for "significant" projects to submit their dialects to a part of the MAVLink repository which is outside of the standard, but maintained to a level that ID clashes are avoided, and possibly some basic style, language, and such things. I believe that the available message and command ID range allows for this.
> 
> And I'd like to get the storm32.xml dialect being admitted to that part of the repository.
> 
> Cheers, Olli
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/mavlink/mavlink/pull/1537
> 
> Commit Summary
> 
> storm32.xml dialect
> File Changes
> 
> A message_definitions/v1.0/storm32.xml (495)
> Patch Links:
> 
> https://github.com/mavlink/mavlink/pull/1537.patch
> https://github.com/mavlink/mavlink/pull/1537.diff
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

olliw42 commented:
@LorenzMeier 

I noted that you plan to remove "legacy" dialects, and I certainly do understand the motivation, and would do that myself if that were my repository.

I was also not expecting this here to be easy, so I also understand that point, and I'm not shocked to death to hear that. Yet, I've decided to put it up. I consider it important enough to at least give it a try.

I think an appealing aspect of the proposal I made, even though I have not explicitely express it, is that such a "outside-of-standard" folder would give you guys much less headaches. One for instance could explicitly establish a rule which says that MAVLink can at any time remove a dialect, e.g., when it appears to MAVLink that it went stale, or if MAVLink wants to grab the IDs themselves, or whatever.

In fact, if you think about it, if there had been such a folder before some of the dialects you are going to clean up now would have had been placed there and wouldn't be an issue in the sense of spilling the standard.
 
@auturgy 

glad to hear that you consider me an active contributor to the MAVLink development. I guess I would however argue that it should be possible to have such dialects been admitted also if this would not be so, but if the project which uses them is of some "significance" or "longer value". Defining what this means is however certainly the difficult part.

1. I fully agree. I in fact just copied what I have in my git repo, without paying attention to cleaning it up to a level it could be merged right away. This is because I was actually expecting that it wouldn't be just go through in a second, and who knows what else one should change in the xml once the rules are settled. So, sorry for the clutter. I will ensure it looks clean once we are there.

The underlying issue would IMHO disappear kind of by itself if such non-standard dialects would be somewhere inside the MAVLink repo.

2. Dito. This sort of clutter definitely would have to go, and replaced by MAVlink's approaches.

RespawnDespair commented:
I know my voice does not carry much weight since I'm not an official MAVLink contributor, but as on of the developers for the Open.HD project I second @olliw42 motion. We are currently reworking our current MAVLink dialect and we experience the same pain as described. So for me personally and I think from a lot of other devs out there this idea gets a +1.

hamishwillee commented:
I was hoping that all the work on the Gimbal protocol essentially would mean that there would be no need for a separate gimbal dialect (more precisely, I was hoping that we would be able to eventually separate out our APIs more finely, and a storm gimbal dialect would include the gimbal API and possibly have a  few extensions).

My understanding is that you want to have the storm dialect in mavlink to make it easy to share for people who want to use this protocol, and to reserve the ranges that you use. Is that correct? So you'd want this included in all.xml. 
- Is the desire that this eventually goes into common/xml? 
- How do people use this? Do they have to rebuild a custom version of the flight stack with it in?

olliw42 commented:
> My understanding is that you want to have the storm dialect in mavlink to make it easy to share for people who want to use this protocol, and to reserve the ranges that you use. Is that correct?

easy to share to people: not the intention.
There are really so many possibilities to share and I really do not want to misuse mavlink as my dropbox. That's not my concern.

reserve ranges: nearly.
I would not ask for some official policy to make ranges reserved. In some way all is needed is a centrally located list of which IDs are used by which dialect, so that the dialect maintainers can see that and respect them if they want to do so. I rather think it should be self-organized, and not yet another burden on the mavlink shoulders.

Also, I would not think there needs to be a policy that mavlink has to respect these ID ranges, i.e., that mavlink at any time just can chose for itself whatever it wants if it determines to need it. It would be kind though if mavlink would not do that on purpose, i.e., would so to say have an eye on these ranges avoid accidental collision.

> So you'd want this included in all.xml.

I haven't really thought about this, but I guess I rather would say no. I guess I would see all.xml as part of the official standard and if I were you I probably would not want to have it overly spilled. But it would be of course an option if you guys would think it is useful. 

> Is the desire that this eventually goes into common/xml?

no, not by construction.
As always, ideas may inspire and someone of you may be inspired by what is in these dialects and suggest to carry it over to common.xml, but it is not the idea that with there is a implied desire to eventually go into it. It in fact may happen that a dialect collides with a micro service/protocol of the standard. I'm afraid I have to say that the storm32 dialect would be an example of that, as it includes an alternative to the standard gimbal protocol. 

> Do they have to rebuild a custom version of the flight stack with it in?

yes, they may have to rebuild a custom version of the flight stack.
This is however highly dependent on the flight stack and what the dialect is about. For example, if the dialect concerns functions which are unrelated to the flight stack, nothing would have to be done. An example could be OpenHD. For example, if the dialect has messages which do not concern the flight controller but only would have to be routed by the flight controller to the component, but the flight stack doesn't have a proper router, then one would have to rebuild the flight stack with these messages included so that it can route these messages too. An example would be storm32 together with ArduPilot. Obviously, if the dialect has functions which need to be part of the flight stack, a custom version is needed.

> How do people use this?

that's indeed a crucial aspect, which I probably haven't pointed out enough. 

just assume that someone for whatever project wants to include both the OpenHD dialect and the STorM32 dialect (e.g. a XyzToMavlink converter), but does not want to do it the silly and hard-to-maintain way by copying files by hand ...  but do it e.g. using git submodules ... it's a kind of nightmare then ... now even assume that OpenHD wants to include the STorM32 dialect and STorM32 wants to include the OpenHD dialect ... oh my gosh ... have not even started to wrap my head around this ...

if these two xml's would be located somewhere in mavlink, one just would need a mavlink fork as usual

julianoes commented:
@olliw42 we lost you on the call. Did zoom crash or your internet die?

LorenzMeier commented:
@RespawnDespair Given that video links are a general thing that a lot of people use, have you considered contributing some of the messages to the standard? It is not just about managing your message IDs, but also about the ability of 3rd parties to integrate easily. The over-arching goal is to get as much "plug and play" as feasible, like the USB standard is achieving. If you plug in a keyboard it just works, you don't need to install a driver (which is == a dialect).

USB / Bluetooth and others have over time adopted more common implementations over time and this is to the benefit of users.

We have a lot of video related functionality already in through the camera protocol: http://mavlink.io/en/services/camera.html

LorenzMeier commented:
@olliw42 The discussion here is still open, we didn't come to a conclusion as we were really just still understanding your needs - do you have the chance to join the next dev call?

olliw42 commented:
@julianoes 
I'm not sure what exactly happened, but I indeed got kicked out ... SORRY for that!

@LorenzMeier 
SORRY to you too. Thx for being open for the matter. 

I assume the next call is 23rd Dez.? I certainly can arrange to join then, if this is considered useful.

Concerning storm32.xml specifically, all the arguments you both put forward I totally agree with. I would not have dared to raise this PR if it would be STorM32 only, and if it were only me - I get my issues solved for myself.

However, talking about standardization, I think one has to acknowledge the fact that this is not always an easy path, and for many smaller projects out of question. Yet, I think that it is a good and vital and healthy thing for MAVLink then also smaller projects decide to chose MAVLink as communication backbone. Not because they would immediately result in the next big contribution to the standard, but as a resource of creativity.

Anyway, I feel a bit like having opened a can of worms, or even box of pandora. If you feel so too and think it's just taking time away, then please do not hesitate to close this PR - I will understand and accept. :)

hamishwillee commented:
Thanks for attending [Devcall](https://github.com/mavlink/mavlink/wiki/20201209-Dev-Meeting) to discuss. End result was "we're not convinced this fits into the direction MAVLink is moving re standardisation etc". We should however not make independent development hard, so we will do what is needed to prevent id collision etc.

hamishwillee commented:
@olliw42 I closed this without reading your post immediately above https://github.com/mavlink/mavlink/pull/1537#issuecomment-741628676 - many apologies.

You didn't get a chance to finish making your points. We ARE having the meeting on 23rd, and you are more than welcome to attend and continue the discussion.

olliw42 commented:
@hamishwillee 

after the meeting, from the arguments which were raised, I had the clear feeling that you were not convinced. My first impulse was similar to yours to close this, and I was a bit struck by surprise by Lorenz' above post saying it is still open, and thus didn't. I VERY much appreciate that, as well as your response. 

However, the process should not be kept open just for the sake of giving me a chance to finish my arguments. That's very polite, but I actually DO have the feeling that you guys have given me a fair chance! And I think my arguments are sufficiently layed out here.

As regards standardization, as said, that's certainly the right thing to do for you - and I sincerely hope that in ten years from now you are this multi-billion dollar company :) - but my opinion is simply that this should not be used as argument against the flock of smaller projects. 

You can't stop people from doing what they want to do, you only can try to embrace and steer them, and ideally benefit from what they do. I deeply believe in that.

Before I dropped out I heard Lorenz explaining what dialects were historically for, but not was came next, which presumably was about what dialects are supposed to be for now. This made me realize that I indeed may have quite misunderstood what MAVLink's idea of dialects is. I may misunderstand the current draft of the governance policy, but it might also be that it is not actually yet totally clear or is somewhat missunderstandable in it's current form. 

I guess a clarification from you folks about the role of dialects, not only for standardization and stakeholders but also for external projects interested in using MAVLink for their needs, would be very helpful. And if it is not yet sharp in your mind, it could be helpful if you could sharpen the idea.

It was probably not very smart of me to just PR the xml without a better indication of what I think the structure finally could be, so I've changed the PR, and have created now a subfolder "nonstandard-dialects" in which to host this kind of xml. This may not be the best solution, and I have not yet tested if it actually works with the current code generator, but it is along the lines of how I would think of it.

And again, if you think your first impulse was appropriate and that this is largely just wasting your time, pl don't hesitate to close, I'll understand.

:)

olliw42 commented:
in preparation for a possible discussion of the topic in the next dev call, I carefully reread the governance policy draft (https://docs.google.com/document/d/1Bb0qF_ldeqGpy3OqtECclrKm00TB1rjZNXCmRopgmio/edit) and the previous meeting notes (https://github.com/mavlink/mavlink/wiki/20201209-Dev-Meeting), and conclude that it IMHO does NOT make much sense to discuss the topic at this point in time.

@hamishwillee: MANY thx for putting it into the next 23.12. dev call, but I would suggest to remove it if you don't disagree.

Even though some statements in the governance policy  make me believe that dialects are supported, e.g. 

"MAVLink dialects are XML files that define protocol- and vendor-specific messages, enums and commands, comply with the MAVLink architecture, but are not part of the formal MAVLink Standard. ... MAVLink Dialects are managed by their owners, within the message space assigned by the MAVLink Maintainers via all.xml.",

it appears to come with relatively strict conditions, as detailed in the "Eligibility for addition of a dialect to the central repository". The last call's meeting notes are saying

"We don't want to support dialects within mavlink unless they have lots of momentum and are evolving towards a standard, but we don't want to prevent people from interworking with the standard. Upshot, we can work with reserving ranges etc."

which as much as I read it raises the bar quite a bit further. With these conditions, the storm32 dialect would be disqualified in actually several points, and as much as I can see also the OpenHD dialect. Hence my conclusion :).

auturgy commented:
We need to rework the draft governance doc in that case.
You’re welcome to PR against ArduPilot/mavlink in the meantime (as is open.HD).


> On 21 Dec 2020, at 7:02 am, olliw42 <notifications@github.com> wrote:
> 
> ﻿
> in preparation for a possible discussion of the topic in the next dev call, I carefully reread the governance policy draft (https://docs.google.com/document/d/1Bb0qF_ldeqGpy3OqtECclrKm00TB1rjZNXCmRopgmio/edit) and the previous meeting notes (https://github.com/mavlink/mavlink/wiki/20201209-Dev-Meeting), and conclude that it IMHO does NOT make much sense to discuss the topic at this point in time.
> 
> @hamishwillee: MANY thx for putting it into the next 23.12. dev call, but I would suggest to remove it if you don't disagree.
> 
> Even though some statements in the governance policy make me believe that dialects are supported, e.g.
> 
> "MAVLink dialects are XML files that define protocol- and vendor-specific messages, enums and commands, comply with the MAVLink architecture, but are not part of the formal MAVLink Standard. ... MAVLink Dialects are managed by their owners, within the message space assigned by the MAVLink Maintainers via all.xml.",
> 
> it appears to come with relatively strict conditions, as detailed in the "Eligibility for addition of a dialect to the central repository". The last call's meeting notes are saying
> 
> "We don't want to support dialects within mavlink unless they have lots of momentum and are evolving towards a standard, but we don't want to prevent people from interworking with the standard. Upshot, we can work with reserving ranges etc."
> 
> which as much as I read it raises the bar quite a bit further. With these conditions, the storm32 dialect would be disqualified in actually several points, and as much as I can see also the OpenHD dialect. Hence my conclusion :).
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

LorenzMeier commented:
@olliw42 Next action point is to continue the conversation on the dev call.

@RespawnDespair Have you seen my post from two weeks ago? I would love to discuss what messages you would like to contribute (either here or on the dev call)

olliw42 commented:
ok, I will be happy to show up

however, admittedly, with somewhat mixed feelings, statements like "what messages you would like to contribute" may hint that the points of view are not very close, but I will look forward to the discussion :)

olliw42 commented:
THANK you guys for the very nice dev call, and positive response. 

Whenever you feel I should do a particular thing, with e.g. this PR but also e.g. with the code generator(s) or whatever else, pl just tell and I'll do my best to bring it forward.

:)

auturgy commented:
The relative path mavgen pr is merged.

Regards,

James 

> On 23 Dec 2020, at 6:24 pm, olliw42 <notifications@github.com> wrote:
> 
> ﻿
> THANK you guys for the very nice dev call, and positive response.
> 
> Whenever you feel I should do a particular thing, with e.g. this PR but also e.g. with the code generator(s) or whatever else, pl just tell and I'll do my best to bring it forward.
> 
> :)
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

olliw42 commented:
I've seen that. MANY thx!

hamishwillee commented:
@olliw42 Lorenz showed up shortly after you left the meeting. 

We agreed that a `mavlink/external` directory can be added for these kinds of non-autopilot dialects. Ultimately we're hoping that most people who do this will come to understand that there are significant benefits to joining the standard itself.

Can you please:
- update the PR to move the files to such a directory.
- Add a comment near the top of your dialect with some contact mechanism - e.g. email or whatever approach you prefer.
- Add a comment near the top of your dialect indicating the ranges you have used for messages/commands
- Include your file in all.xml 
- Add a readme indicating that ranges can be requested by attending the devcall. 

Hope this makes sense. We can iterate.

olliw42 commented:
fantastic !

I'll do, but probably not today :)

have all a nice Christmas

olliw42 commented:
hey folks,

so, this was a bit more complicated than I thought. The generator had to be updated, which took me a while. Also, pushing is giving me now troubles, thx to git workflow and authentication issues ... need to see how I get this right (did I say already that I hate git?). Anyways:

> * update the PR to move the files to such a directory.

I moved the storm32.xml to a folder **_external/dialects_** (and not just external as discussed before).
Reason is that it considered it possible that it may happen that in some distant future you may want to add other "external" stuff, such as code generator, or whatever else. And external/dialects just makes better reference to dialects...

> * Add a comment near the top of your dialect with some contact mechanism - e.g. email or whatever approach you prefer.
> * Add a comment near the top of your dialect indicating the ranges you have used for messages/commands

done

I hope that's about ok, pl feel free to suggest changes

> * Include your file in all.xml

done

For all this to work , one needs however these PRs 
https://github.com/ArduPilot/pymavlink/pull/494
https://github.com/ArduPilot/pymavlink/commit/c655e2312ec8dc0de9aa63ffe67f2b0fa43f38ab

> * Add a readme indicating that ranges can be requested by attending the devcall.

done

this however I would like to see just as an a suggestion, i.e., should (or needs to) be discussed. Not only language.-wise but also content-wise.

I have added e.g. a statement that IDs should be above 60000, and would be typically reserved in blocks of 50. I'm not sure this makes sense, or goes with your governance philosophy (e.g. hoping to get messages transferred to common.xml).

**My tests so far**

I have tested that all.xml produces exactly the same code (modulo of course of the expected code changes like MAV_CMD enum or this IDX thing), for three scenarios: 1. all.xml without storm32.xml, 2. all.xml with storm32.xml from external folder, 3. all.xml with storm32.xml from normal xxx/v1.0 folder. 

I have tested that storm32.xml produces exactly the same code (modulo the IDX thing) if generated form external or from the default xxx/1.0.

I have tested that the STorM32 firmware still compiles with using the external mechanism.

What I have NOT yet done are functional test, i.e., if the freshly compiled STorM32 firmware works, if works for my ArduPilot and OpenTx forks, etc.. Given that the code looks ok to me I don't expect surprises, but I should do these testes of course.

**Dependencies/limitations**

I've mentioned the two PRs to pymavlink before.

Dialects are distinguished by the file names only. That is, one can't place files with equal name in different folders, they would be seen as to be one and the same, and this probably gives issues. I think that's an ok limitation.

I have no idea why the checks fail, presumably a formatting issue with storm32.xml which I can't identify. I have tried using scripts/format_xml.sh but couldn't figure how to get it to work (I'm on win, and I'm a linux nut). Having said this, I would love to add a statement to the above README.md that one should use scripts/format_xml.sh, along with a short comment of how to use it ...

olliw42 commented:
hey folks

both required PRs to pymavlink have been merged now ... 
(it's strange that the other PR https://github.com/ArduPilot/pymavlink/pull/495 appears to not have been even considered, not easy to understand management principles at work here, but anyway)

so, I've made my first attempt to update a git submodule in a PR ... I hope I got this right (locally it appears to works LOL)

thus, from my perspective, the PR is good now. 
Let's see what your responses are :)

and: HAPPY NEW YEAR (it only can get better, right :))

olliw42 commented:
I think I got the formatting of the storm32.xml right now, at least it passes format_xml.sh now. The checks still fail unfortunately ... something goes wrong with JS-NextGen as it seems ... I guess I would need some help here.

olliw42 commented:
ping
@hamishwillee

hamishwillee commented:
Hi @olliw42 
FYI only working a couple of days per week at the moment, and following holidays. Will put this on my list to look at tomorrow.

hamishwillee commented on message_definitions/v1.0/all.xml:
```diff
@@ -17,5 +17,7 @@
   <include>ualberta.xml</include>
   <include>uAvionix.xml</include>
+  <!-- external dialects -->
+  <include>../../external/dialects/storm32.xml</include>
```
Perhaps we should ask for them to list the ranges here? This way we can just point them to this file and suggest they allocate the next 50 message and command ids in the list?

olliw42 replied:
I think your underlying idea for putting the ranges also in all.xml is to have a "complete overview" in one place. I think this is a good idea.

hamishwillee commented:
@olliw42 Looking good. I reworded the readme a bit to add links to files. Also added note making it clear that we'd rather people engage with the standard :-).

I also suggested they submit request for a range in the initial PR. I have implied rather than stated that the dev call is not scary - it is just a way for us to engage better.

Looks like tests are passing - the  JS-NextGen thingy was a test case issue fixed in Pymavlink last week.

@auturgy @LorenzMeier @julianoes Do you have anything to add? Looks to me like Ollie has done as discussed.

olliw42 commented:
first, many thx for going intensely over it. I had to laugh a bit about "rewording a bit", you certainly improved the text quite a bit ...

I guess I would write "You are asked to attend the  Dev call (once)..." instead of "You may need to attend...". I would have no idea which conditions would trigger the "may need".

One aspect, which I tried to somehow bring across, albeit admittedly somewhat clumsily, and which I am missing a bit in the current wording, is to reach a welcoming warm hand to the external projects. While you of course want folks to get engaged with the standard, I think you even more want folks to conisder submitting their dialect in the first place. So it should be IMHO encouraging, and maybe encourage even those who would normally think that their voice doesn't carry much weight. I mean, if after a few years there would still be just one dialect in that folder, then it would not have been worth the effort and certainly would have failed the goals, wouldn't it?

E.g., I guess I rather would write something along the lines "We strongly recommend that you work with the standard rather than using your own dialect and this mechanism wherever possible (there are significant benefits in terms of compatibility and adoptability when using the standard definitions), but also encourage you to submit your dialect through this mechanism when using a own dialect you consider crucial for your project."

yes, your last pymavlink update made the fails go away - magically.

hamishwillee commented:
Hi @olliw42 Cool! 
- I've added note that all.xml should include the ranges too
- I have changed this to say that you will generally be invited to attend dev call for final approval. The change from "must" is because sometimes we might not want to invite people. 
- My take from the devcall was that we don't want this to be very welcoming - we'd much rather people work with the standard. So I intended this to be a little discouraging. I take your point though - no point offering a gateway to working with the standard if no one will take it.

I'll ask for devcall "sign off" this evening.

olliw42 commented:
Hey @hamishwillee 

many thx!

as regards en- vs dis-couraging, I guess we can see it and evaluate in a year or so if that mechanism is worth it :) And if not, e.g. because it is then still me, one probably could want to remove it with a "hey, was worth a try" :)

olliw42 commented:
sorry, question: In light of the "60000 + 50" "rule", should I modify id ranges to 60000-60049 or to 62000-62049 or it's ok as is?

hamishwillee commented:
@olliw42 Pretty much the dev call agrees with you - let's get this in and evaluate whether it is useful after we've tried it. Thanks very much for all your work on this.

Yes, can you please modify the ranges - everyone will copy you, so starting from the bottom makes sense. If this is a significant problem for you then we could update docs to start from 62000 instead. Also please add the ranges used to the all.xml.

Then I am happy to merge. 

Thanks!

olliw42 commented:
> Yes, can you please modify the ranges - everyone will copy you, so starting from the bottom makes sense. If this is a significant problem for you then we could update docs to start from 62000 instead. 

it certainly would be quite a bit easier to not change, but it is absolutely worth the effort ! So done :)

> Also please add the ranges used to the all.xml.

:)

MANY THX!

hamishwillee commented:
Thanks Ollie, merged! Yee har.

olliw42 commented:
I've learned it as: Yiiiiiiiiiiiiiihhhhhhhaaaaaa

THANKS, very cool :)
